File: go/cf.go
package cf

import (
	"compress/flate"
	"encoding/binary"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

const (
	signature      = 0x7FFFFFFF
	imageHeaderLen = 4 + 4 + 4 + 4     // uint32 + uint32 + uint32 + uint32
	pageHeaderLen  = 2 + 9 + 9 + 9 + 2 // CRLF + hex8byte_ + hex8byte_ + hex8byte_ + CRLF
	rowHeaderLen   = 8 + 8 + 4         // datetime + datetime + attr
	unknown4       = 4
)

// Reader ...
type Reader struct {
	src string
	pos int
	len int
	unp bool
}

type header struct {
	fullSize int
	pageSize int
	nextPage int
}

// OpenFile ...
func OpenFile(path string) Reader {
	r := Reader{}
	data, err := ioutil.ReadFile(path)
	if err != nil {
		panic(err)
	}
	r.src = string(data)
	r.len = len(r.src)
	r.unp = true
	return r
}

// OpenString ...
func OpenString(data []byte) Reader {
	r := Reader{}
	r.src = string(data)
	r.len = len(r.src)
	r.unp = false
	return r
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func (r *Reader) setpos(pos int) {
	r.pos = pos
	if r.pos >= r.len {
		panic("r.pos >= len")
	}
}

func (r *Reader) read(count int) string {
	pos := r.pos
	r.pos += count
	if r.pos > r.len {
		panic("r.pos > len")
	}
	return r.src[pos:r.pos]
}

func (r *Reader) skip(count int) {
	r.pos += count
	if r.pos >= r.len {
		panic("r.pos >= len")
	}
}

func (r *Reader) readPageHeader() header {
	raw := r.read(pageHeaderLen)
	fullSize, err := strconv.ParseInt(raw[2:10], 16, 64)
	check(err)
	pageSize, err := strconv.ParseInt(raw[11:19], 16, 64)
	check(err)
	nextPage, err := strconv.ParseInt(raw[20:28], 16, 64)
	check(err)
	return header{int(fullSize), int(pageSize), int(nextPage)}
}

func (r *Reader) readRowHeaderID() string {
	pageHeader := r.readPageHeader()
	r.skip(rowHeaderLen)
	return r.read(pageHeader.fullSize - rowHeaderLen - unknown4)
}

func (r *Reader) readRowBody() string {
	pageHeader := r.readPageHeader()
	fullSize := pageHeader.fullSize
	size := min(fullSize, pageHeader.pageSize)
	t := []string{}
	t = append(t, r.read(size))
	fullSize = fullSize - size
	for pageHeader.nextPage != signature {
		r.setpos(pageHeader.nextPage)
		pageHeader = r.readPageHeader()
		size := min(fullSize, pageHeader.pageSize)
		t = append(t, r.read(size))
		fullSize = fullSize - size
	}
	return strings.Join(t, "")
}

func (r *Reader) readPointers() []uint32 {
	r.skip(imageHeaderLen)
	t := []uint32{}
	pageHeader := r.readPageHeader()
	fullSize := pageHeader.fullSize
	readSize := 0
	for pageHeader.nextPage != signature {
		readSize = readSize + pageHeader.pageSize
		for i := 0; i < pageHeader.pageSize; i += 4 {
			var p uint32
			err := binary.Read(strings.NewReader(r.read(4)), binary.LittleEndian, &p)
			check(err)
			t = append(t, p)
		}
		r.setpos(pageHeader.nextPage)
		pageHeader = r.readPageHeader()
	}
	for i := 0; i < fullSize-readSize; i += 4 {
		var p uint32
		err := binary.Read(strings.NewReader(r.read(4)), binary.LittleEndian, &p)
		check(err)
		t = append(t, p)
	}
	return t
}

// NewRowsReader ...
func (r *Reader) NewRowsReader() func() (*string, []byte) {
	i := 0
	pointers := r.readPointers()
	return func() (*string, []byte) {
		if i < len(pointers) {
			r.setpos(int(pointers[i]))
			id := strings.Replace(r.readRowHeaderID(), "\x00", "", -1)
			r.setpos(int(pointers[i+1]))
			raw := r.readRowBody()
			var body []byte
			var err error
			if r.unp {
				body, err = ioutil.ReadAll(flate.NewReader(strings.NewReader(raw)))
				check(err)
			} else {
				body = []byte(raw)
			}
			i += 3
			return &id, body
		}
		return nil, nil
	}
}

// Leaf ...
type Leaf struct {
	beg, end, next int
}

// Tree ...
type Tree struct {
	list []Leaf
	src  string
	pos  int
	len  int
}

// Init ...
func (t *Tree) Init(src string) {
	t.list = make([]Leaf, 0, 256)
	t.src = src
	t.pos = 0
	t.len = len(src)
}

func (t *Tree) Read(path ...int) string {
	pos := 0
	for _, index := range path {
		for i := 1; i < index; i++ {
			pos = t.list[pos].next
		}
		pos++
	}
	elem := t.list[pos-1]
	return t.src[elem.beg:elem.end]
}

func (t *Tree) String() string {
	return t.src
}

// Print ...
func (t *Tree) Print() {
	for _, v := range t.list {
		println(v.beg, v.end)
	}
}

// Parse ...
func (t *Tree) Parse() {
	n := 0
loop:
	for t.pos < t.len {
		switch t.src[t.pos] {
		case '{':
			t.list = append(t.list, Leaf{t.pos, 0, 0})
			p := len(t.list) - 1
			t.pos++
			t.Parse()
			t.list[p].end = t.pos + 1
			t.list[p].next = len(t.list)
			n = -1
		case '}':
			break loop
		case '"':
			n++
			t.pos++
			for t.pos < t.len && t.src[t.pos] != '"' {
				n++
				t.pos++
			}
			n++
		case ',':
			if n >= 0 {
				t.list = append(t.list, Leaf{t.pos - n, t.pos, len(t.list) + 1})
			}
			n = 0
		default:
			n++
		}
		t.pos++
	}
	if n >= 0 {
		t.list = append(t.list, Leaf{t.pos - n, t.pos, len(t.list) + 1})
	}
}

// Dir ...
type Dir map[string]interface{}

// Load ...
func Load(path string) Dir {
	dir := make(Dir)
	f := OpenFile(os.Args[1])
	readRow := f.NewRowsReader()
	id, body := readRow()
	for id != nil {
		if string(body[0:4]) == "\xFF\xFF\xFF\x7F" {
			s := OpenString(body)
			readSubRow := s.NewRowsReader()
			subdir := make(Dir)
			subid, body := readSubRow()
			for subid != nil {
				t := Tree{}
				t.Init(string(body))
				subdir[*subid] = &t
				subid, body = readSubRow()
			}
			dir[*id] = subdir
		} else {
			t := Tree{}
			t.Init(string(body))
			dir[*id] = &t
		}

		id, body = readRow()
	}
	return dir
}

File: lua/cf.lua
local END  = 0x7FFFFFFF
local BOM  = "\xEF\xBB\xBF"
local CRLF = "\x0D\x0A"
local ImageHeaderLen = 4 + 4 + 4 + 4     -- uint32 + uint32 + uint32 + uint32
local PageHeaderLen  = 2 + 9 + 9 + 9 + 2 -- CRLF + hex8byte_ + hex8byte_ + hex8byte_ + CRLF
local RowHeaderLen   = 8 + 8 + 4         -- datetime + datetime + attr
local Unknown4       = 4

local null = setmetatable({}, {
    __tostring = function ()
        return "null"
    end
})

local Module = {}

local Reader = {}
Reader.__index = Reader

local FileReader = setmetatable({}, Reader)
FileReader.__index = FileReader
FileReader.file = null
FileReader.fpos = 0
FileReader.flen = 0
FileReader.offset = 0
FileReader.buflen = 0
FileReader.buffer = null

function Module.FileReader(file, buflen)
    local this = setmetatable({}, FileReader)
    this.file   = file
    this.offset = 0
    this.fpos   = 0
    this.flen   = file:seek("end")
    this.buflen = buflen or 256 * 1024
    file:seek("set")
    if this.flen < this.buflen then
        this.buflen = this.flen
    end
    this.buffer = file:read(this.buflen)
    return this
end

function FileReader:Seek(newpos)
    assert(newpos <= self.flen)
    self.offset = newpos % self.buflen
    newpos = newpos - self.offset
    if self.fpos ~= newpos then
        self.fpos = self.file:seek("set", newpos);
        self.buffer = self.file:read(math.min(self.buflen, self.flen - newpos))
    end
end

function FileReader:Read(len)
    local t = {}
    local tail = self.buflen - self.offset
    if len > tail then
        t[#t+1] = self.buffer:sub(self.offset + 1, self.buflen)
        self:Seek(self.fpos + self.buflen)
        len = len - tail
    end
    while len >= self.buflen do
        t[#t+1] = self.buffer
        self:Seek(self.fpos + self.buflen)
        len = len - self.buflen
    end
    if self.buffer then
        local newoffset = self.offset + len
        t[#t+1] = self.buffer:sub(self.offset + 1, newoffset)
        self.offset = newoffset
    end
    return table.concat(t)
end

local StringReader = setmetatable({}, Reader)
StringReader.__index = StringReader
StringReader.src = null
StringReader.len = 0
StringReader.pos = 0

function Module.StringReader(src)
    local this = setmetatable({}, StringReader)
    this.src = src
    this.len = #src
    this.pos = 0
    return this
end

function StringReader:Seek(newpos)
    assert(newpos <= self.len)
    self.pos = newpos
end

function StringReader:Read(len)
    local start = self.pos + 1
    self.pos = self.pos + len
    if self.pos <= self.len then
        return self.src:sub(start, self.pos)
    else
        return ""
    end
end

local function getUInt32(b0, b1, b2, b3)
    return b3 * 256^3 + b2 * 256^2 + b1 * 256 + b0
end

function Reader:ReadImageHeader()
    local s = self:Read(ImageHeaderLen)
    return {
        PageSize  = getUInt32(s:byte(5,  8)),
        Revision  = getUInt32(s:byte(9, 12)),
        Unknown   = getUInt32(s:byte(13, 16))
    }
end

function Reader:ReadPageHeader()
    local s = self:Read(PageHeaderLen)
    assert(s:sub( 1,  2) == CRLF)
    assert(s:sub(30, 31) == CRLF)
    return {
        FullSize = tonumber(s:sub( 3, 10), 16),
        PageSize = tonumber(s:sub(12, 19), 16),
        NextPage = tonumber(s:sub(21, 28), 16)
    }
end

function Reader:ReadRowHeader()
    local pageHeader = self:ReadPageHeader()
    assert(pageHeader.NextPage == END)
    local s = self:Read(RowHeaderLen)
    return {
        Creation   = s:sub(1,  8),
        Modified   = s:sub(9, 16),
        Attributes = getUInt32(s:byte(17, 20)),
        ID = self:Read(pageHeader.FullSize - RowHeaderLen - Unknown4)
    }
end

function Reader:ReadRowBody()
    local pageHeader = self:ReadPageHeader()
    local fullSize = pageHeader.FullSize
    local size = math.min(fullSize, pageHeader.PageSize)
    local t = {self:Read(size)}
    fullSize = fullSize - size
    while pageHeader.NextPage ~= END do
        self:Seek(pageHeader.NextPage)
        pageHeader = self:ReadPageHeader()
        size = math.min(fullSize, pageHeader.PageSize)
        t[#t+1] = self:Read(size)
        fullSize = fullSize - size
    end
    assert(fullSize == 0)
    return table.concat(t)
end

function Reader:ReadPointers()
    local t = {}
	local pageHeader = self:ReadPageHeader()
	local fullSize = pageHeader.FullSize
	local readSize = 0
    local size, data
    while pageHeader.NextPage ~= END do
        size = pageHeader.PageSize
        readSize = readSize + size
        data = self:Read(size)
        for i = 1, size, 4 do
			t[#t+1] = getUInt32(data:byte(i,  i + 3))
        end
		self:Seek(pageHeader.NextPage)
		pageHeader = self:ReadPageHeader()
    end
    size = fullSize - readSize
    data = self:Read(size)
    for i = 1, size, 4 do
        t[#t+1] = getUInt32(data:byte(i,  i + 3))
    end
    t.rd = self
	return t
end

local function iter(t, i)
    local id, body, header
    local pos = t[i]
    if pos == nil then
        return nil
    end
    assert(t[i+2] == END)
    t.rd:Seek(pos)
    header = t.rd:ReadRowHeader()
    id = string.gsub(header.ID, "%z", "")
    t.rd:Seek(t[i+1])
    body = t.rd:ReadRowBody()
    i = i + 3
    return i, id, body, body:sub(1, 3) ~= BOM, header
end

function Module.ReadImage(rd)
    local header = rd:ReadImageHeader()
    local pointers = rd:ReadPointers()
    return {
        Header = header,
        Pointers = pointers,
        Rows = function ()
            return iter, pointers, 1
        end,
        List = function ()
            local t = {}
            local ID
            for i = 1, #pointers, 3 do
                rd:Seek(pointers[i])
                ID = string.gsub(rd:ReadRowHeader().ID, "%z", "")
                t[ID] = pointers[i+1]
            end
            return t
        end
    }
end

function Module.Parse(s, beg)
    beg = beg or 1
    local pos = s:find('{', beg, true)
    if not pos then return nil end
    local qt, lb, rb = ('"{}'):byte(1,3)
    local skip, b = false, 0
    local function parse()
        beg = pos + 1
        pos = s:find('["{},]', pos + 1)
        local tree = {}
        while pos do
            b = s:byte(pos)
            if b == qt then
                skip = not skip
            elseif not skip then
                if b == lb then
                    tree[#tree+1] = assert(parse())
                    pos = pos + 1
                    beg = pos + 1
                elseif b == rb then
                    if beg < pos then
                        tree[#tree+1] = s:sub(beg, pos - 1)
                    end
                    return tree
                else -- ','
                    tree[#tree+1] = s:sub(beg, pos - 1)
                    beg = pos + 1
                end
            end
            pos = s:find('["{},]', pos + 1)
        end
        return {}
    end
    return parse()
end

return Module